dex/src/state.rs

```diff
@@ -59,6 +60,150 @@ pub enum AccountFlag {
+            Market::V2(v2) => v2.deref(),
+        }
+    }
+}
+
+impl<'a> DerefMut for Market<'a> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        match self {
+            Market::V1(v1) => v1.deref_mut(),
+            Market::V2(v2) => v2.deref_mut(),
+        }
+    }
+}
+
+impl<'a> Market<'a> {
+    #[inline]
+    pub fn load(market_account: &'a AccountInfo, program_id: &Pubkey) -> DexResult<Self> {
+        if market_account.try_data_len()? == MarketState::LEN {
+            Ok(Market::V1(MarketState::load(market_account, program_id)?))
+        } else {
+            Ok(Market::V2(MarketStateV2::load(market_account, program_id)?))
+        }
+    }
+
+    pub fn open_orders_authority(&self) -> Option<&Pubkey> {
+        match &self {
+            Market::V1(_) => None,
+            Market::V2(state) => Some(&state.open_orders_authority),
+        }
+    }
+
+    pub fn load_orders_mut(
+        &self,
+        orders_account: &'a AccountInfo,
+        owner_account: Option<&AccountInfo>,
+        program_id: &Pubkey,
+        rent: Option<Rent>,
+        open_orders_authority: Option<account_parser::SignerAccount>,
+    ) -> DexResult<RefMut<'a, OpenOrders>> {
+        check_assert_eq!(orders_account.owner, program_id)?;
+        let mut open_orders: RefMut<'a, OpenOrders>;
+
+        let open_orders_data_len = orders_account.data_len();
+        let open_orders_lamports = orders_account.lamports();
+        let (_, data) = strip_header::<[u8; 0], u8>(orders_account, true)?;
+        open_orders = RefMut::map(data, |data| from_bytes_mut(data));
+
+        if open_orders.account_flags == 0 {
+            let oo_authority = open_orders_authority.map(|a| a.inner().key);
+            if oo_authority != self.open_orders_authority() {
+                return Err(DexErrorCode::InvalidMarketAuthority.into());
+            }
+
+            let rent = rent.ok_or(DexErrorCode::RentNotProvided)?;
+            let owner_account = owner_account.ok_or(DexErrorCode::OwnerAccountNotProvided)?;
+            if !rent.is_exempt(open_orders_lamports, open_orders_data_len) {
+                return Err(DexErrorCode::OrdersNotRentExempt)?;
+            }
+            open_orders.init(
+                &identity(self.own_address),
+                &owner_account.key.to_aligned_bytes(),
+            )?;
+        }
+
+        open_orders.check_flags()?;
+        check_assert_eq!(identity(open_orders.market), identity(self.own_address))
+            .map_err(|_| DexErrorCode::WrongOrdersAccount)?;
+        if let Some(owner) = owner_account {
+            check_assert_eq!(&identity(open_orders.owner), &owner.key.to_aligned_bytes())
+                .map_err(|_| DexErrorCode::WrongOrdersAccount)?;
+        }
+
+        Ok(open_orders)
+    }
+}
+
+#[derive(Copy, Clone)]
+#[cfg_attr(target_endian = "little", derive(Debug))]
+#[repr(packed)]
+pub struct MarketStateV2 {
+    pub inner: MarketState,
+    pub open_orders_authority: Pubkey,
+}
```

Comment:
Might be worth placing some padding bytes at the end of the account for future upgrades.

Reply:
More fields will be added in subsequent PRs such as https://github.com/project-serum/serum-dex/pull/157, so we can consider this in those changes.